src/dynamics/mpas/dyn_comp.F90

```diff
@@ -620,22 +618,37 @@ subroutine set_mpas_state_rho_theta()
 
                 ! Piecewise integrate hypsometric equation to derive `p_mid_col(1)`.
                 ! The formulation used here is exact.
-                p_mid_col(1) = p_sfc(i) * &
-                    exp(-0.5_kind_r8 * (zgrid(2, i) - zgrid(1, i)) * &
-                    constant_g / (constant_rd * tm_mid_col(1)) * (1.0_kind_r8 + qv_mid_col(1)))
+                ! p_mid_col(1) = p_sfc(i) * &
+                !     exp(-0.5_kind_r8 * (zgrid(2, i) - zgrid(1, i)) * &
+                !     constant_g / (constant_rd * tm_mid_col(1)) * (1.0_kind_r8 + qv_mid_col(1)))
+                p_mid_col(1) = p_by_hypsometric_equation( &
+                    p_sfc(i), &
+                    zgrid(1, i), &
+                    tm_mid_col(1) / (1.0_kind_r8 + qv_mid_col(1)), &
+                    0.5_kind_r8 * (zgrid(2, i) + zgrid(1, i)))
 
                 ! Piecewise integrate hypsometric equation to derive subsequent `p_mid_col(k)`.
                 ! The formulation used here is exact.
                 do k = 2, pver
-                    p_mid_col(k) = p_mid_col(k - 1) * &
-                        exp(-0.5_kind_r8 * (zgrid(k    , i) - zgrid(k - 1, i)) * &
-                        constant_g / (constant_rd * tm_mid_col(k - 1)) * (1.0_kind_r8 + qv_mid_col(k - 1))) * &
-                        exp(-0.5_kind_r8 * (zgrid(k + 1, i) - zgrid(k    , i)) * &
-                        constant_g / (constant_rd * tm_mid_col(k    )) * (1.0_kind_r8 + qv_mid_col(k    )))
+                    ! p_mid_col(k) = p_mid_col(k - 1) * &
+                    !     exp(-0.5_kind_r8 * (zgrid(k    , i) - zgrid(k - 1, i)) * &
+                    !     constant_g / (constant_rd * tm_mid_col(k - 1)) * (1.0_kind_r8 + qv_mid_col(k - 1))) * &
+                    !     exp(-0.5_kind_r8 * (zgrid(k + 1, i) - zgrid(k    , i)) * &
+                    !     constant_g / (constant_rd * tm_mid_col(k    )) * (1.0_kind_r8 + qv_mid_col(k    )))
+                    p_mid_col(k) = p_by_hypsometric_equation( &
+                        p_by_hypsometric_equation( &
+                            p_mid_col(k - 1), &
+                            0.5_kind_r8 * (zgrid(k, i) + zgrid(k - 1, i)), &
+                            tm_mid_col(k - 1) / (1.0_kind_r8 + qv_mid_col(k - 1)), &
+                            zgrid(k, i)), &
+                        zgrid(k, i), &
+                        tm_mid_col(k) / (1.0_kind_r8 + qv_mid_col(k)), &
+                        0.5_kind_r8 * (zgrid(k + 1, i) + zgrid(k, i)))
                 end do
 
                 rho(:, i) = p_mid_col(:) / (constant_rd * tm_mid_col(:))
-                theta(:, i) = t_mid(:, i) * ((constant_p0 / p_mid_col(:)) ** (constant_rd / constant_cpd))
+                ! theta(:, i) = t_mid(:, i) * ((constant_p0 / p_mid_col(:)) ** (constant_rd / constant_cpd))
```

Comment:
I might remove the commented-out code here.

Reply:
I pushed https://github.com/ESCOMP/CAM-SIMA/pull/267/commits/3ca360bb44c2935ae58afce307c82a2829cc51b6 to remove the commented code.